### PR TITLE
[quant][fx] Support some default ops in the native backend config

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #74603
* #74601
* __->__ #74600

Summary:
Following https://github.com/pytorch/pytorch/pull/74210, this PR adds the support for some ops
using the DefaultNodeQuantizeHandler in the backend_config_dict defintion for pytorch native backend

TODO: There is still a few ops we didn't handle with backend_config_dict path: gelu and softmax, need to discuss if we still need them, if so we can change the test
to use backend_config_dict and remove the DefaultNodeQuantizeHandler after that

Test Plan:
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestFXNumericSuiteCoreAPIs
python test/test_quantization.py TestFXGraphMatcher

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35071437](https://our.internmc.facebook.com/intern/diff/D35071437)

### CODE REVIEW

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35071437).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35071437).

File: torch/ao/quantization/fx/backend_config/native.py [hunk 1]
```diff
@@ -19,6 +29,25 @@ def get_native_backend_config_dict():
         # optional, output activation dtype
         "output_dtype": torch.quint8
     }
+
+    # input_activation: quint8, output_activation: quint8
```
Comment (andrewor14) [inline torch/ao/quantization/fx/backend_config/native.py:33]: nit: make these comments consistent across different dtype configs? Also I feel some of these comments are actually not needed, like this one doesn't give much additional information IMO

File: torch/ao/quantization/fx/backend_config/native.py [hunk 2]
```diff
@@ -19,6 +29,25 @@ def get_native_backend_config_dict():
...
+
+    default_op_fp16_dtype_config = {
+        # optional, input activation dtype
+        "input_dtype": torch.float16,
+        # optional, weight dtype
+        "weight_dtype": torch.float16,
+        # optional, output activation dtype
+        "output_dtype": torch.float16,
+    }
+    # END dtype configs
```
Comment (andrewor14) [inline torch/ao/quantization/fx/backend_config/native.py:49]: nit: I wonder if we should define these configs outside of `get_native_backend_config_dict` to make it easier to read, maybe in a future PR, otherwise this function will grow really large

Comment (andrewor14) [reply torch/ao/quantization/fx/backend_config/native.py:49]: Such that `get_native_backend_config_dict` will only look like this:
```
def get_native_backend_config_dict():
    """
    ...
    """
    return {
        # optional
        "name": "native",
        "configs": [
            linear_module_config,
            *default_op_configs,
            *default_op_fp16_configs,
             ...
        ],
    }
```
What do you think?

**Review by andrewor14 — APPROVED** [review]

File: torch/ao/quantization/fx/backend_config/native.py (hunk 1 above)
Comment (jerryzh168) [reply torch/ao/quantization/fx/backend_config/native.py:33]: sounds good, I can remove these

File: torch/ao/quantization/fx/backend_config/native.py (hunk 2 above)
Comment (jerryzh168) [reply torch/ao/quantization/fx/backend_config/native.py:49]: yeah mean define everything outside? I think we may want to hide some of the implementation details, don't really want to expose everything in native.py, but we can expose the dtype configs, it might be useful for user as well

Comment (jerryzh168) [reply torch/ao/quantization/fx/backend_config/native.py:49]: I guess if we prefix these things with `_` it could work

File: test/quantization/fx/test_numeric_suite_fx.py [hunk 3]
```diff
@@ -681,9 +680,9 @@ def _op_is_unmatchable(op):
                 self.assertTrue(
                     _op_in_base_sets_of_related_ops(base_op),
                     f"{base_op} not in sets of related ops")
-            else:
+            elif not _op_in_base_sets_of_related_ops(base_op):
```
Comment (jerryzh168) [inline test/quantization/fx/test_numeric_suite_fx.py:683]: @vkuzo this is the change I made to the test, please take a look

Comment (jerryzh168) [reply test/quantization/fx/test_numeric_suite_fx.py:683]: basically we don't explicitly check the quantize handler type in this branch, but just directly check if op is in the related ops set

File: test/quantization/fx/test_numeric_suite_fx.py [hunk 4]
```diff
@@ -682,8 +678,14 @@ def _op_is_unmatchable(op):
                     _op_in_base_sets_of_related_ops(base_op),
                     f"{base_op} not in sets of related ops")
             else:
-                raise AssertionError(
-                    f"handing for {qhandler_cls} not implemented")
+                # torch.sum does not have quantized equivalents
```
Comment (jerryzh168) [inline test/quantization/fx/test_numeric_suite_fx.py:681]: here is the change

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35071437).

File: test/quantization/fx/test_numeric_suite_fx.py [hunk 5]
```diff
@@ -682,8 +678,14 @@ def _op_is_unmatchable(op):
...
+                if base_op == torch.sum:
+                    continue
+                # didn't match explicit quantize handler class, we can check if the
+                # operator is in the related op set directly
+                if not _op_in_base_sets_of_related_ops(base_op):
```
Comment (jerryzh168) [inline test/quantization/fx/test_numeric_suite_fx.py:686]: @vkuzo please take a look at this change

Comment (vkuzo) [reply test/quantization/fx/test_numeric_suite_fx.py:686]: looks like this is now testing a smaller set of ops, could we keep the set of ops under test the same?  This would mean querying the structure you added (`get_native_backend_config_dict`) and taking the union of ops in there and ops in quantize handlers.

Comment (jerryzh168) [reply test/quantization/fx/test_numeric_suite_fx.py:686]: not sure I follow, are you talking `get_default_quant_patterns` not returning the patterns in `get_native_backend_config_dict`?

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35071437).

File: test/quantization/fx/test_numeric_suite_fx.py (hunk 5 above)
Comment (vkuzo) [reply test/quantization/fx/test_numeric_suite_fx.py:700]: oh, my bad, I didn't realize that you added this in `get_all_quant_patterns`, sorry. Look good!

Comment (jerryzh168) [reply test/quantization/fx/test_numeric_suite_fx.py:700]: oh I corrected it later after your comment, although some how the new patterns does appear as well, probably because we are modifying the patterns dictionary inplace..

Comment (jerryzh168) [reply test/quantization/fx/test_numeric_suite_fx.py:700]: I'll change get_quant_patterns to return a copy, instead of the original dictionary

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35071437).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35071437).
